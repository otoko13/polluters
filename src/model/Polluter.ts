export interface IPolluterArgs {
  rank: number;
  name: string;
  longitude: number;
  latitude: number;
  globalEmissions: string;
  ownership: string;
  hq: string;
  ceo: string | undefined;
  ceoAnnualPay: string | undefined;
  fossilFuelProduction: string | undefined;
  investmentInRenewables: string | undefined;
  globalEmissions1965to2017: string | undefined;
  projectedIncreaseInProduction2018to2030: string | undefined;
  projectedEmissions2018to2030: string | undefined;
  environmentalDisaster: string | undefined;
  environmentalScandal: string | undefined;
  futureProjects: string | undefined;
  revenue: string | undefined;
}

export enum EOwnership{
  State,
  Shareholders,
  StateAndShareholders,
  Unknown,
}

export interface IPolluter extends Polluter {}

const NO_INFORMATTON = 'No information';

class Polluter {
  readonly name: string;
  readonly rank: number;
  readonly longitude: number;
  readonly latitude: number;
  readonly globalEmissions: string;
  readonly ownership: EOwnership;
  readonly hq: string;
  readonly ceo: string;
  readonly annualCeoPay: number | undefined;
  readonly fossilFuelProduction: string;
  readonly investmentInRenewables: string;
  readonly globalEmissions1965to2017: string;
  readonly projectedIncreaseInProduction2018to2030: string;
  readonly projectedEmissions2018to2030: string;
  readonly environmentalDisaster: string | undefined;
  readonly environmentalScandal: string | undefined;
  readonly futureProjects: string;
  readonly revenue: string;

  private static getOwnership(ownership: string) {
    const isState = ownership.toLowerCase().indexOf('state') > -1;
    const isShareholder = ownership.toLowerCase().indexOf('shareholder') > -1;
    if (isState && isShareholder) {
      return EOwnership.StateAndShareholders;
    }
    if (isState) {
      return EOwnership.State;
    }
    if (isShareholder) {
      return EOwnership.Shareholders;
    }
    return EOwnership.Unknown;
  }

  private static getCeoAnnualPay(value: string | undefined) {
    if (!value) {
      return;
    }
    const regEx = /^\$(.+)m(.)*$/gi;
    const payMatch = regEx.exec(value);
    if (!payMatch) {
      return;
    }
    return parseFloat(payMatch[1]);
  }

  constructor(args: IPolluterArgs) {
    this.name = args.name;
    this.rank = args.rank;
    this.longitude = args.longitude;
    this.latitude = args.latitude;
    this.globalEmissions = args.globalEmissions;
    this.hq = args.hq;
    this.ceo = args.ceo || 'Unknown';
    this.fossilFuelProduction = args.fossilFuelProduction || NO_INFORMATTON;
    this.investmentInRenewables = args.investmentInRenewables || NO_INFORMATTON;
    this.globalEmissions1965to2017 = args.globalEmissions1965to2017 || NO_INFORMATTON;
    this.projectedIncreaseInProduction2018to2030 = args.projectedIncreaseInProduction2018to2030 || NO_INFORMATTON;
    this.projectedEmissions2018to2030 = args.projectedEmissions2018to2030 || NO_INFORMATTON;
    this.environmentalDisaster = args.environmentalDisaster;
    this.environmentalScandal = args.environmentalScandal;
    this.futureProjects = args.futureProjects || NO_INFORMATTON;
    this.revenue = args.revenue || NO_INFORMATTON;

    this.annualCeoPay = Polluter.getCeoAnnualPay(args.ceoAnnualPay);
    console.log(this.annualCeoPay);

    this.ownership = Polluter.getOwnership(args.ownership);
  }
}

export default Polluter;
