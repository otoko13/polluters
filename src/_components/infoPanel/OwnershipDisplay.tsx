import React from 'react';
import GavelIcon from '@material-ui/icons/Gavel';
import PeopleAltIcon from '@material-ui/icons/PeopleAlt';
import { EOwnership } from '../../model/Polluter';
import './ownershipDisplay.scss';

export interface IOwnershipDisplayProps {
  ownership: EOwnership;
}

function getOwnershipText(status: EOwnership) {
  switch (status) {
    case EOwnership.Shareholders:
      return 'Shareholders';
    case EOwnership.State:
      return 'State';
    case EOwnership.StateAndShareholders:
      return 'State and shareholders';
    default:
      return 'Unknown';
  }
}

function getOwnershipIcons(status: EOwnership) {
  switch (status) {
    case EOwnership.Shareholders:
      return <span><PeopleAltIcon /></span>;
    case EOwnership.State:
      return <span><GavelIcon /></span>;
    case EOwnership.StateAndShareholders:
      return <span><GavelIcon /> <PeopleAltIcon /></span>;
    default:
      return undefined;
  }
}

const OwnershipDisplay = ({ ownership }: IOwnershipDisplayProps) => (
  <div className="OwnershipDisplay">
    <div className="icons">
      {getOwnershipIcons(ownership)}
    </div>
    <div className="text">
      {getOwnershipText(ownership)}
    </div>
  </div>
);

export default OwnershipDisplay;
