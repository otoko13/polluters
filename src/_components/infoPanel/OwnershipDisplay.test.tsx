import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import OwnershipDisplay from './OwnershipDisplay';
import { EOwnership } from '../../model/Polluter';

describe('snapshots', () => {
  it('should look sensible for state', () => {
    const { container } = render(<OwnershipDisplay ownership={EOwnership.State} />);
    expect(container.firstChild).toMatchSnapshot('State');
  });

  it('should look sensible for shareholders', () => {
    const { container } = render(<OwnershipDisplay ownership={EOwnership.Shareholders} />);
    expect(container.firstChild).toMatchSnapshot('Shareholders');
  });

  it('should look sensible for state and shareholders', () => {
    const { container } = render(<OwnershipDisplay ownership={EOwnership.StateAndShareholders} />);
    expect(container.firstChild).toMatchSnapshot('State and shareholders');
  });
});