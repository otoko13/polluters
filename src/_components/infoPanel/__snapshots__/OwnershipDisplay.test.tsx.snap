// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should look sensible for shareholders: Shareholders 1`] = `
<div
  class="OwnershipDisplay"
>
  <div
    class="icons"
  >
    <span>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.67 13.13C18.04 14.06 19 15.32 19 17v3h4v-3c0-2.18-3.57-3.47-6.33-3.87z"
          fill-rule="evenodd"
        />
        <circle
          cx="9"
          cy="8"
          fill-rule="evenodd"
          r="4"
        />
        <path
          d="M15 12c2.21 0 4-1.79 4-4s-1.79-4-4-4c-.47 0-.91.1-1.33.24C14.5 5.27 15 6.58 15 8s-.5 2.73-1.33 3.76c.42.14.86.24 1.33.24zM9 13c-2.67 0-8 1.34-8 4v3h16v-3c0-2.66-5.33-4-8-4z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
  </div>
  <div
    class="text"
  >
    Shareholders
  </div>
</div>
`;

exports[`snapshots should look sensible for state and shareholders: State and shareholders 1`] = `
<div
  class="OwnershipDisplay"
>
  <div
    class="icons"
  >
    <span>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M5.2496 8.0688l2.83-2.8268 14.134 14.15-2.83 2.8268zM9.4857 3.8272l2.828-2.8288 5.6576 5.656-2.828 2.8288zM.9989 12.3147l2.8284-2.8284L9.484 15.143l-2.8284 2.8284zM1 21h12v2H1z"
        />
      </svg>
       
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.67 13.13C18.04 14.06 19 15.32 19 17v3h4v-3c0-2.18-3.57-3.47-6.33-3.87z"
          fill-rule="evenodd"
        />
        <circle
          cx="9"
          cy="8"
          fill-rule="evenodd"
          r="4"
        />
        <path
          d="M15 12c2.21 0 4-1.79 4-4s-1.79-4-4-4c-.47 0-.91.1-1.33.24C14.5 5.27 15 6.58 15 8s-.5 2.73-1.33 3.76c.42.14.86.24 1.33.24zM9 13c-2.67 0-8 1.34-8 4v3h16v-3c0-2.66-5.33-4-8-4z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
  </div>
  <div
    class="text"
  >
    State and shareholders
  </div>
</div>
`;

exports[`snapshots should look sensible for state: State 1`] = `
<div
  class="OwnershipDisplay"
>
  <div
    class="icons"
  >
    <span>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M5.2496 8.0688l2.83-2.8268 14.134 14.15-2.83 2.8268zM9.4857 3.8272l2.828-2.8288 5.6576 5.656-2.828 2.8288zM.9989 12.3147l2.8284-2.8284L9.484 15.143l-2.8284 2.8284zM1 21h12v2H1z"
        />
      </svg>
    </span>
  </div>
  <div
    class="text"
  >
    State
  </div>
</div>
`;
